crates/bevy_ecs/src/observer/runner.rs

```diff
@@ -302,24 +302,40 @@ impl<E: Event, B: Bundle> Observer<E, B> {
     /// # Safety
     /// The type of the `event` [`ComponentId`] _must_ match the actual value
     /// of the event passed into the observer system.
-    pub unsafe fn with_event(mut self, event: ComponentId) -> Self {
+    pub unsafe fn with_event_unchecked(mut self, event: ComponentId) -> Self {
         self.descriptor.events.push(event);
         self
     }
 }
 
-impl<E: Event, B: Bundle> Component for Observer<E, B> {
+impl<B: Bundle> Observer<DynamicEvent, B> {
+    /// Observe the given `event`. This will cause the [`Observer`] to run whenever an event with the given [`ComponentId`]
+    /// is triggered.
+    ///
+    /// # Note
+    /// As opposed to [`Observer::with_event_unchecked`], this method is safe to use because no pointer casting is performed automatically.
+    /// That is left to the user to do manually.
+    pub fn with_event(self, event: ComponentId) -> Self {
+        // SAFETY: DynamicEvent itself does not perform any unsafe operations (like casting), so this is safe.
+        unsafe { self.with_event_unchecked(event) }
```

Comment:
This safety comment is not entirely clear to me. It doesn't seem to address the safety requirements of the `with_event_unchecked` function:

> ```
> # Safety
> The type of the `event` [`ComponentId`] _must_ match the actual value
> of the event passed into the observer system.
> ```

In your safety comment there's no mention of the `event` `ComponentId` or the type of the actual event that will be passed into the observer system.

I guess this is safe because `DynamicEvent` can "match" any event type? If that's the case I would like to see clearer safety requirements from `with_event_unchecked` that allow for this situation, and a better safety comment here that addresses the requirements from `with_event_unchecked`.